### PR TITLE
Add stable unstable version for jump to date before `v1.6` is fully supported on a homeserver

### PR DESCRIPTION
**Pull request by MadLittleMods**
Add stable unstable version (`org.matrix.msc3030.stable`) for jump to date [before `v1.6` is fully supported on a homeserver](https://github.com/matrix-org/synapse/issues/15089). Discussion for whether we are okay with moving forward with this -> https://github.com/matrix-org/matrix-react-sdk/pull/10398#discussion_r1139847241

Related to https://github.com/vector-im/element-web/issues/24362 but does not solve immediately because Synapse does not supply `org.matrix.msc3030.stable` yet

Also refactored `ServerSupportUnstableFeatureController` to support multiple feature groups where any one of them will enable the setting. All features in a feature group are required. See discussion -> https://github.com/matrix-org/matrix-react-sdk/pull/10398#discussion_r1139850466

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [ ] ~~Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))~~

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

File: src/settings/Settings.tsx [hunk 1]
```diff
@@ -372,8 +372,8 @@ export const SETTINGS: { [setting: string]: ISetting } = {
         controller: new ServerSupportUnstableFeatureController(
             "feature_jump_to_date",
             defaultWatchManager,
-            ["org.matrix.msc3030"],
-            undefined,
+            ["org.matrix.msc3030", "org.matrix.msc3030.stable"],
```
Comment (MadLittleMods) [inline src/settings/Settings.tsx:375]: All of these features are *AND*ed together when we actually want them to be *OR*ed 🤔

Comment (t3chguy) [reply src/settings/Settings.tsx:375]: Guess it should be extended to string[][] where its an OR of AND groups to offer more flexibility

Comment (MadLittleMods) [reply src/settings/Settings.tsx]: > Guess it should be extended to string[][] where its an OR of AND groups to offer more flexibility

Added support for this 👍

File: src/settings/Settings.tsx [hunk 2]
```diff
@@ -372,8 +372,8 @@ export const SETTINGS: { [setting: string]: ISetting } = {
         controller: new ServerSupportUnstableFeatureController(
             "feature_jump_to_date",
             defaultWatchManager,
-            ["org.matrix.msc3030"],
-            undefined,
+            [["org.matrix.msc3030"], ["org.matrix.msc3030.stable"]],
```
Comment (MadLittleMods) [inline src/settings/Settings.tsx:375]: Add stable unstable version (`org.matrix.msc3030.stable`) for jump to date [before `v1.6` is fully supported on a homeserver](https://github.com/matrix-org/synapse/issues/15089).

`org.matrix.msc3030.stable` wasn't included in [MSC3030](https://github.com/matrix-org/matrix-spec-proposals/pull/3030) but it seems like pattern that should have been included like [MSC2285](https://github.com/matrix-org/matrix-spec-proposals/pull/2285) as an example.

Are we okay with moving forward with this?

File: src/settings/controllers/ServerSupportUnstableFeatureController.ts [hunk 3]
```diff
@@ -26,12 +26,21 @@ import SettingsStore from "../SettingsStore";
...
+    // it as a change and updates the watchers.
     private enabled: boolean | undefined;
 
+    /**
+     * Construct a new ServerSupportUnstableFeatureController.
+     *
+     * @param unstableFeatureGroups - If any one of the feature groups is satisfied,
+     * then the setting is considered enabled. A feature group is satisfied if all of
+     * the features in the group are supported (all features in a group are required).
+     */
```
Comment (MadLittleMods) [inline src/settings/controllers/ServerSupportUnstableFeatureController.ts:39]: *As suggested by @t3chguy in https://github.com/matrix-org/matrix-react-sdk/pull/10398#discussion_r1139471244,*

> Guess it should be extended to string[][] where its an OR of AND groups to offer more flexibility

Refactoring `ServerSupportUnstableFeatureController` to support multiple feature groups where any one of them will enable the setting. All features in a feature group are required.

This way having either `org.matrix.msc3030` or `org.matrix.msc3030.stable` will enable the jump to date feature flag with a config of `[["org.matrix.msc3030"], ["org.matrix.msc3030.stable"]]`

If you want to *AND* feature together like before, just add them to the same group: `[["foo", "bar"]]`

File: src/settings/Settings.tsx (hunk 1 above)
Comment (MadLittleMods) [reply src/settings/Settings.tsx]: Noted this change on the diff in this comment -> https://github.com/matrix-org/matrix-react-sdk/pull/10398#discussion_r1139850466

**Review by t3chguy — APPROVED** [review]
Looks great, thanks for extending the controller

**Comment by robintown** [comment]
Related: https://github.com/vector-im/element-web/issues/24362

**Comment by MadLittleMods** [comment]
@robintown Feel free to edit this type of info directly in the description

**Comment by MadLittleMods** [comment]
Thanks for the review @t3chguy and taking a look @robintown 🐮

File: src/settings/Settings.tsx (hunk 2 above)
Comment (MadLittleMods) [reply src/settings/Settings.tsx:375]: Related discussion https://github.com/matrix-org/synapse/pull/15283#discussion_r1140669410

**Pull request merged**